### PR TITLE
chore: Match InfluxQL OG time range behaviour when gap filling

### PR DESCRIPTION
**Pull request by stuartcarnie**
Helps #7600

This PR is additional work towards introducing window function support, which aligns time-range behaviour with InfluxQL OG when gap filling. Most notably, a `Filter` node is added with the the upper bound of time to `≤ now` when a a `GROUP BY TIME` clause is present.

**Other notable changes**

- The `#[allow(deadcode)]` attribute was added to a few functions that will be used with window aggregates in a later PR
- When evaluating a duration expression, `now` is allowed, which addresses compatibility with `GROUP BY TIME(5s, now())`, where `now` is specified as the offset.

### CODE REVIEW

File: influxdb_influxql_parser/src/expression/conditional.rs [hunk 1]
```diff
@@ -260,10 +260,7 @@ pub fn parse_conditional_expression(input: &str) -> Result<ConditionalExpression
     let mut i: &str = input;
 
     // Consume whitespace from the input
-    i = match ws0(i) {
-        Ok((i1, _)) => i1,
-        _ => unreachable!("ws0 is infallible"),
-    };
+    (i, _) = ws0(i).expect("ws0 is infallible");
```
Comment (stuartcarnie) [inline influxdb_influxql_parser/src/expression/conditional.rs:263]: This change was made to match the pattern used elsewhere in the InfluxQL parser

File: influxdb_influxql_parser/src/time_range.rs [hunk 2]
```diff
@@ -390,12 +390,9 @@ impl TimeRange {
...
 /// Simplifies an InfluxQL duration `expr` to a nanosecond interval represented as an `i64`.
-pub fn duration_expr_to_nanoseconds(expr: &Expr) -> Result<i64, ExprError> {
-    let ctx = ReduceContext::default();
-    match reduce_expr(&ctx, expr)? {
-        Expr::Literal(Literal::Duration(v)) => Ok(*v),
-        Expr::Literal(Literal::Float(v)) => Ok(v as i64),
-        Expr::Literal(Literal::Integer(v)) => Ok(v),
+pub fn duration_expr_to_nanoseconds(ctx: &ReduceContext, expr: &Expr) -> Result<i64, ExprError> {
+    match reduce_time_expr(ctx, expr)? {
+        Expr::Literal(Literal::Timestamp(v)) => Ok(v.timestamp_nanos()),
```
Comment (stuartcarnie) [inline influxdb_influxql_parser/src/time_range.rs:395]: This ensures `now()` is also evaluated when simplifying InfluxQL expressions that are expected to be a duration.

File: iox_query_influxql/src/plan/planner.rs [hunk 3]
```diff
@@ -176,17 +199,175 @@ impl<'a> Context<'a> {
...
             group_by: select.group_by.as_ref(),
             fill: select.fill,
+            interval: select.interval,
+            extra_intervals: select.extra_intervals,
             root_group_by_tags: self.root_group_by_tags,
         }
     }
 
+    /// Return a [`Expr::Sort`] expression for the `time` column.
+    #[allow(dead_code)]
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:219]: Will be used in a future PR for #7600

File: iox_query_influxql/src/plan/planner.rs [hunk 4]
```diff
@@ -176,17 +199,175 @@ impl<'a> Context<'a> {
...
+                OrderByClause::Descending => false,
+            },
+            false,
+        )
+    }
+
+    /// Returns true if the current context has an extended
+    /// time range to provide leading data for window functions
+    /// to produce the result for the first window.
+    #[allow(dead_code)]
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:233]: Will be used in a future PR for #7600

File: iox_query_influxql/src/plan/planner.rs [hunk 5]
```diff
@@ -176,17 +199,175 @@ impl<'a> Context<'a> {
...
+    ///     if opt.Ascending {
+    ///         opt.StartTime -= int64(opt.Interval.Duration)
+    /// ```
+    ///
+    /// [1]: https://github.com/influxdata/influxdb/blob/f365bb7e3a9c5e227dbf66d84adf674d3d127176/query/compile.go#L592-L594
+    /// [2]: https://github.com/influxdata/influxdb/blob/f365bb7e3a9c5e227dbf66d84adf674d3d127176/query/compile.go#L1153-L1158
+    /// [3]: https://github.com/influxdata/influxdb/blob/f365bb7e3a9c5e227dbf66d84adf674d3d127176/query/compile.go#L1172-L1173
+    /// [4]: https://github.com/influxdata/influxdb/blob/f365bb7e3a9c5e227dbf66d84adf674d3d127176/query/compile.go#L1198
+    /// [5]: https://github.com/influxdata/influxdb/blob/f365bb7e3a9c5e227dbf66d84adf674d3d127176/query/select.go#L259-L261
+    #[allow(dead_code)]
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:343]: Will be used in a future PR for #7600

File: iox_query_influxql/src/plan/planner.rs [hunk 6]
```diff
@@ -2962,18 +3140,20 @@ mod test {
...
-                    GapFill: groupBy=[[time]], aggr=[[selector_last(cpu.usage_idle,cpu.time)]], time_column=time, stride=IntervalMonthDayNano("5000000000"), range=Unbounded..Excluded(now()) [time:Timestamp(Nanosecond, None);N, selector_last(cpu.usage_idle,cpu.time):Struct([Field { name: "value", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "time", data_type: Timestamp(Nanosecond, None), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }]);N]
+                    GapFill: groupBy=[[time]], aggr=[[selector_last(cpu.usage_idle,cpu.time)]], time_column=time, stride=IntervalMonthDayNano("5000000000"), range=Unbounded..Included(TimestampNanosecond(1672531200000000000, None)) [time:Timestamp(Nanosecond, None);N, selector_last(cpu.usage_idle,cpu.time):Struct([Field { name: "value", data_type: Float64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "time", data_type: Timestamp(Nanosecond, None), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }]);N]
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner.rs:3143]: Because we are resolving `now` in the InfluxQL planner, it is now represented as a literal value.

Comment (wolffcm) [reply iox_query_influxql/src/plan/planner.rs:3143]: I imagine that once you have support for the windowing functions like `moving_average`,  you'll need to add an additional filter after the `WindowExec` node that will remove the "extra" windows?

**Review by wolffcm — APPROVED** [review]
Looks good!

File: iox_query_influxql/src/plan/planner.rs (hunk 6 above)
Comment (stuartcarnie) [reply iox_query_influxql/src/plan/planner.rs:3143]: Gosh, it's so confusing and it turns out it depends on a few other factors…  I'm still in the process of codifying those, as the behaviour is different when its a subquery vs the root too…

**Pull request merged**